### ISSUE TITLE
Add ES2 support to webrender

### ISSUE THREAD

**Issue by mmatyas** [servo/webrender#3701]
Currently [webrender requires OpenGL 3.2](https://github.com/servo/servo/blob/691367745ad27ef79d67e4dc9be679eb969fdd96/ports/glutin/window.rs#L170), which is not supported by many embedded devices, as they use ES 2.0 (GLSL 1.2). What kind of 3.2 specific features Servo uses? Is it possible to add ES2 support?

**Comment by jrmuizel** [comment]
Yes, that's the plan.

**Comment by nical** [comment]
> What kind of 3.2 specific features Servo uses?

Off the top of my head, the features we use that are not universally supported on ES2 devices:

 - texture arrays
 - instancing
 - texture fetches in vertex shaders
 - 32 bits float textures

**Comment by dsx724** [comment]
Is there an hour/funding estimate to work around the missing features and implement GLES2? We still have quite a bit of GLES2 infrastructure and currently have to use Chrome which has other issues.

**Comment by jrmuizel** [comment]
What hardware are you running on? The recommend approach to running WebRender on ES2 hardware is with Software WebRender (SWGL) and OpenGL ES2 compositing. This is what we do on Android phones that don't support ES3

**Comment by dsx724** [comment]
Is there any particular runtime flags or preference settings we need to enable? We are running Debian 12 with both weston and mutter/Gnome. The hardware is ARM Cortex-A53 with ARM Mali-450 and we have thousands deployed around the world.

**Comment by jrmuizel** [comment]
Ok, so you're getting Software WebRender without ES2 compositing. We don't enable ES2 compositing by default on Linux so that's expected. Is there a situation or particular piece of content that has bad performance? Are you using Firefox from the debian package?

It might be possible to force it on by setting `layers.acceleration.force-enabled` to true in about:config and restarting.
What do you get for HW_COMPOSITING/OPENGL_COMPOSITING/WEBRENDER if you do that?

**Comment by dsx724** [comment]
Yes, Firefox from the Debian package. We're seeing around 5x the performance with Chromium and WebGL. Scrolling any page and the difference is very significant. Chrome is pretty smooth where as Firefox hangs for hundreds of milliseconds.

```
WebGL creation failed: 
* tryNativeGL (FEATURE_FAILURE_EGL_CREATE)
* Exhausted GL driver options. (FEATURE_FAILURE_WEBGL_EXHAUSTED_DRIVERS)
```

**Comment by jrmuizel** [comment]
Are you able to find out why EGL Create is failing here?
https://searchfox.org/mozilla-central/rev/606cdd34860e42611fceba025314864cd293a43b/gfx/gl/GLContextProviderEGL.cpp#723

**Comment by anarsoul** [comment]
@dsx724 chromium (and chrome) is able to use GLES2 for rendering and compositing, that's why it's smooth. Firefox reverts to software rendering which is much slower, especially on Cortex A53.

@jrmuizel does desktop Firefox try to use GLES2 if it sees that desktop GL is available? Lima driver exposes desktop GL 2.1 (it's not actually compliant, but works fine with most apps) and GLES 2.0. Firefox used to be smooth on Lima until they dropped GL compositing.

**Comment by jrmuizel** [comment]
@anarsoul by default we don't try to use GLES2 or desktop GL on desktop Linux, only software. It may be possible to force either of them but I'm not sure of the details around doing so.

**Comment by X-m7** [comment]
For what it's worth the PinePhone (non-Pro) is also stuck on OpenGL ES 2.0 due to its Mali-400MP2 GPU, and it can still be bought new today. It's also not particularly powerful in the CPU department with its 4 Cortex-A53 cores at 1.15GHz either.
